analyzers/src/SonarAnalyzer.CSharp/Rules/SillyMathematicalComparison.cs

```diff
@@ -0,0 +1,89 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public sealed class SillyMathematicalComparison : SonarDiagnosticAnalyzer
+    {
+        private const string DiagnosticId = "S2198";
+        private const string MathComparisonMessage = "Comparison to this constant is useless; the constant is outside the range of type '{0}'";
+
+        private static readonly DiagnosticDescriptor MathComparisonRule = DescriptorFactory.Create(DiagnosticId, MathComparisonMessage);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(MathComparisonRule);
+
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterNodeAction(
+                CheckComparisonOutOfRange,
+                CSharpFacade.Instance.SyntaxKind.ComparisonKinds);
+
+        private static void CheckComparisonOutOfRange(SonarSyntaxNodeReportingContext context)
+        {
+            if (TryGetConstantValue(context.SemanticModel, (BinaryExpressionSyntax)context.Node, out var constant, out var other)
+               && context.SemanticModel.GetTypeInfo(other).Type is { } typeSymbolOfOther
+               && TryGetRange(typeSymbolOfOther) is { } range
+               && (constant < range.Min || constant > range.Max))
```

Comment:
NIT: Could be made a method on the `ValueRange` type
```suggestion
               && range.IsOutOfRange(constant)
```